File: packages/fjs/src/project/engine.ts
// JS engine flavor selection (spec 091, reshaped by spec 105).
//
// The flutter_fjs package ships BOTH engines prebuilt under abi/ and the
// platform builds pick one themselves: android/build.gradle and the
// podspecs read FJS_JS_ENGINE (environment first, then the host's
// --dart-define). This module resolves the flavor, exports it into the
// environment every later `flutter` child inherits, and runs the package's
// Dart runner (bin/engine.dart) for the two things a build file cannot do
// on its own: copy the flavor into ohos/libs (HARs only package libs/) and
// make the HOST re-run pod install after a switch. `dart run` resolves
// flutter_fjs through the host's package_config, so pure Flutter hosts run
// the very same command:
//
//   dart run flutter_fjs:engine quickjs
import { spawnSync } from 'node:child_process';
import path from 'node:path';

export type JsEngine = 'primjs' | 'quickjs';

const ENGINES: JsEngine[] = ['primjs', 'quickjs'];

/** Engine id each flavor bakes into .fjsbundle headers (fjs_engine_id()). */
export const ENGINE_IDS: Record<JsEngine, string> = {
  primjs: 'primjs-4.1.1',
  quickjs: 'quickjs-ng-0.9.0',
};

/** --js-engine value, falling back to FJS_JS_ENGINE, then the default.
 * An explicit but unknown value throws with the accepted list. */
export function resolveJsEngine(explicit?: string): JsEngine {
  const raw = explicit ?? process.env.FJS_JS_ENGINE;
  if (!raw) return 'primjs';
  if ((ENGINES as string[]).includes(raw)) return raw as JsEngine;
  throw new Error(`unknown js engine '${raw}' — expected one of: ${ENGINES.join(', ')}`);
}

/** Selects [engine] for the Flutter build that follows. Sets
 * `FJS_JS_ENGINE` in this process's environment (the gradle file and the
 * podspecs read it; every `flutter` spawned afterwards inherits it), then
 * runs the flutter_fjs engine runner from [flutterDir] (see the header).
 *
 * The runner is not optional for a non-default flavor — ohos libs and the
 * host's pods depend on it — so a missing `dart` or a failing runner throws
 * then. For the default flavor android/ios/macos need nothing from it, so
 * the failure is reported once and the command goes on; it is never
 * silent, because the next build may still carry a previous switch.
 * `debugger` is accepted for the callers' sake and has no effect since
 * spec 105: release builds leave the debugger out at the build layer. */
export function materializeJsEngine(
  engine: JsEngine,
  opts: {
    flutterDir?: string;
    explicit?: boolean;
    debugger?: boolean;
    /** test seam */
    spawn?: typeof spawnSync;
    /** test seam */
    warn?: (line: string) => void;
  } = {},
): void {
  process.env.FJS_JS_ENGINE = engine;
  const cwd = opts.flutterDir ? path.resolve(opts.flutterDir) : process.cwd();
  const spawn = opts.spawn ?? spawnSync;
  const warn = opts.warn ?? ((line: string) => console.warn(line));
  const r = spawn('dart', ['run', 'flutter_fjs:engine', engine, '--host', cwd], {
    cwd,
    stdio: 'pipe',
  });
  const tolerable = engine === 'primjs' && !opts.explicit;
  if (r.error) {
    const why =
      `cannot run 'dart' (${r.error.message}) — the engine runner ships inside ` +
      'flutter_fjs and needs the Dart SDK on PATH';
    if (tolerable) {
      warn(
        `fjs: engine runner skipped — ${why}. android/ios/macos still build primjs; ` +
          'an ohos host or a host switched away from primjs earlier needs the runner.',
      );
      return;
    }
    throw new Error(`--js-engine ${engine}: ${why}`);
  }
  const out = `${r.stdout ?? ''}${r.stderr ?? ''}`;
  if (r.status !== 0) {
    if (tolerable) {
      warn(`fjs: engine runner failed — ${out.trim().split('\n')[0] ?? 'no output'}`);
      return;
    }
    throw new Error(`fjs engine runner failed:\n${out}`);
  }
  // the runner's own line may share stdout with dart's build-hook chatter
  const line = out.match(/fjs: engine flavor[^\r\n]*/)?.[0];
  if (line) console.log(line);
}

File: packages/flutter_fjs/bin/engine.dart
// Engine-flavor runner (spec 091, reshaped by spec 105).
//
//   dart run flutter_fjs:engine quickjs     # from any host that depends on
//   dart run flutter_fjs:engine primjs      # this package
//
// Android, iOS and macOS no longer need it to switch engines: their build
// files pick the flavor themselves (android/build.gradle points jniLibs at
// abi/<flavor>/android; the podspecs vendor <platform>/abi/<flavor>/), so
// this package's directory is never written for them. That matters because
// for most hosts the directory is the shared pub cache: copying a flavor
// over it (what spec 091 did) let two projects clobber each other, let a
// release build of one strip the debugger out of another's debug build,
// and made `pub publish` ship whatever the last run left on disk.
//
// What is left here:
//
//  - ohos: a HAR packages native libraries from the module's libs/ only, so
//    the flavor still has to be copied into ohos/libs. The copy compares
//    bytes first (the default primjs set ships in place, so the common case
//    writes nothing) and is refused when this package sits in the pub cache
//    — use a path dependency to run a non-default flavor on ohos. Skipped
//    entirely for hosts without an ohos/ directory.
//  - the HOST's CocoaPods state: the podspecs choose the flavor at pod
//    install, and flutter skips pod install when nothing it tracks changed.
//    The last flavor is recorded in the host's .dart_tool; on a change the
//    host Podfile is touched and the host's Xcode xcframework caches are
//    dropped, so the next build re-installs and relinks. Host files only.
//
// --no-debugger is accepted for compatibility and does nothing: release and
// profile builds already leave the debugger out at the build layer (gradle
// exclude, the linker never pulling the unreferenced archive, the ohos
// buildModeBinder filter).
import 'dart:convert';
import 'dart:io';
import 'dart:isolate';

const _engineIds = <String, String>{
  'primjs': 'primjs-4.1.1',
  'quickjs': 'quickjs-ng-0.9.0',
};

void main(List<String> args) {
  final engine = args.isNotEmpty ? args.first : (Platform.environment['FJS_JS_ENGINE'] ?? '');
  if (engine != 'primjs' && engine != 'quickjs') {
    stderr.writeln(
        "usage: dart run flutter_fjs:engine <primjs|quickjs> [--host <dir>] — "
        "got '${args.isEmpty ? '' : args.first}'");
    exit(64);
  }
  // --host <dir>: the host project. Defaults to the working directory, which
  // is where a pure Flutter host runs this from.
  var hostDir = Directory.current.path;
  for (var i = 1; i < args.length; i++) {
    if (args[i] == '--host' && i < args.length - 1) {
      hostDir = args[i + 1].replaceAll(RegExp(r'/+$'), '');
    }
  }

  final root = _findPackageRoot();
  final ohos = _selectOhos(root, hostDir, engine);
  final pods = _markHostFlavor(hostDir, engine);
  stdout.writeln('fjs: engine flavor ${_engineIds[engine]} — android/ios/macos '
      'select it at build time; ohos libs $ohos; host pods $pods');
}

/// Makes ohos/libs hold [engine]'s libraries. Returns a word for the
/// summary line; exits when a copy is needed but not allowed.
String _selectOhos(String root, String host, String engine) {
  if (!Directory('$host/ohos').existsSync()) return 'not used by this host';
  final src = Directory('$root/abi/$engine/ohos/arm64-v8a');
  if (!src.existsSync()) {
    stderr.writeln('fjs: engine flavor incomplete — missing ${src.path}');
    exit(66);
  }
  final dest = Directory('$root/ohos/libs/arm64-v8a');
  final wanted = {for (final f in src.listSync().whereType<File>()) _name(f.path): f};
  final present = dest.existsSync()
      ? {for (final f in dest.listSync().whereType<File>()) _name(f.path): f}
      : <String, File>{};
  final same = wanted.length == present.length &&
      wanted.entries.every((e) => present[e.key] != null && _sameBytes(e.value, present[e.key]!));
  if (same) return 'already $engine';
  final cache = _pubCacheRoot();
  if (cache != null && _isWithin(root, cache)) {
    stderr.writeln(
        'fjs: the $engine engine on ohos needs its libraries copied into flutter_fjs/ohos/libs,\n'
        'but this flutter_fjs is in the shared pub cache ($root) and is not modified.\n'
        'Depend on flutter_fjs through a path dependency to use a non-default engine on ohos.');
    exit(69);
  }
  dest.createSync(recursive: true);
  // quickjs has no debugger module; a primjs one left behind would bind
  // LEPUS_* symbols the quickjs engine does not have
  for (final name in present.keys) {
    if (!wanted.containsKey(name)) present[name]!.deleteSync();
  }
  for (final e in wanted.entries) {
    e.value.copySync('${dest.path}/${e.key}');
  }
  return 'switched to $engine';
}

/// Records [engine] in the host's .dart_tool and, when it changed, makes the
/// next Xcode build re-run pod install and relink. Host files only.
String _markHostFlavor(String host, String engine) {
  // v2 (spec 112): hosts that ran spec 105's runner hold a v1 stamp that
  // may already name the new flavor while their build still links the old
  // one (the cache drop never matched), so the v1 stamp is not trusted.
  final stamp = File('$host/.dart_tool/flutter_fjs/engine_flavor.v2');
  // no stamp: we cannot know what the last build linked — invalidate once
  // (one pod install and a relink) rather than guess
  final previous = stamp.existsSync() ? stamp.readAsStringSync().trim() : null;
  if (previous == engine) return 'unchanged';
  stamp.parent.createSync(recursive: true);
  stamp.writeAsStringSync('$engine\n');
  // flutter skips pod install when the plugin set is unchanged — a newer
  // Podfile makes it re-run, and the podspec re-evaluates the flavor
  for (final platform in ['ios', 'macos']) {
    final podfile = File('$host/$platform/Podfile');
    if (podfile.existsSync()) podfile.setLastModifiedSync(DateTime.now());
  }
  _dropXcodeCaches(Directory('$host/build'));
  return 'invalidated (${previous ?? 'unknown'} → $engine)';
}

/// Deletes, anywhere under the host's build dir, the copy of the engine
/// slice Xcode extracted (`XCFrameworkIntermediates/flutter_fjs`) and the
/// linked plugin framework (`flutter_fjs/flutter_fjs.framework`).
///
/// Both flavors ship a `libfjs.a` under the same name, and the CocoaPods
/// copy phase re-runs only when its input is NEWER than its output — a
/// quickjs archive checked out earlier than the last primjs copy looked
/// up to date, and a switch kept linking primjs (spec 112). The search is
/// recursive because Flutter nests these a few levels down:
///   build/ios/Debug-iphonesimulator/XCFrameworkIntermediates/flutter_fjs
///   build/macos/Build/Products/Debug/XCFrameworkIntermediates/flutter_fjs
/// Spec 091 looked only at build/<one level>/…, which never matched; it
/// did not show then because every switch re-copied files with fresh
/// timestamps. Keep in step with tool/test/xcode_cache_paths_check.mjs.
void _dropXcodeCaches(Directory build, [int depth = 0]) {
  if (depth > 6 || !build.existsSync()) return;
  for (final entry in build.listSync(followLinks: false).whereType<Directory>()) {
    final name = _name(entry.path);
    if (name == 'XCFrameworkIntermediates') {
      final stale = Directory('${entry.path}/flutter_fjs');
      if (stale.existsSync()) stale.deleteSync(recursive: true);
      continue;
    }
    if (name == 'flutter_fjs') {
      final product = Directory('${entry.path}/flutter_fjs.framework');
      if (product.existsSync()) product.deleteSync(recursive: true);
      continue;
    }
    // bundles never contain either; skipping them keeps the walk cheap
    if (name.endsWith('.app') || name.endsWith('.framework') || name.endsWith('.dSYM')) continue;
    _dropXcodeCaches(entry, depth + 1);
  }
}

String? _pubCacheRoot() {
  final env = Platform.environment;
  final explicit = env['PUB_CACHE'];
  if (explicit != null && explicit.isNotEmpty) return explicit;
  if (Platform.isWindows) {
    final local = env['LOCALAPPDATA'];
    return local == null ? null : '$local\\Pub\\Cache';
  }
  final home = env['HOME'];
  return home == null ? null : '$home/.pub-cache';
}

bool _isWithin(String path, String dir) {
  final p = Directory(path).absolute.uri.normalizePath().path;
  var d = Directory(dir).absolute.uri.normalizePath().path;
  if (!d.endsWith('/')) d = '$d/';
  return p.startsWith(d);
}

String _name(String p) => p.split(Platform.pathSeparator).last;

bool _sameBytes(File a, File b) {
  if (a.lengthSync() != b.lengthSync()) return false;
  final da = a.readAsBytesSync(), db = b.readAsBytesSync();
  for (var i = 0; i < da.length; i++) {
    if (da[i] != db[i]) return false;
  }
  return true;
}

/// Locates this package's own checkout. `dart run` executes a kernel
/// snapshot, so Platform.script points into .dart_tool — resolve the
/// package root from the host's package_config instead (the same mechanism
/// `dart run` itself used to find us), walking up for good measure.
String _findPackageRoot() {
  final resolved = Isolate.resolvePackageUriSync(Uri.parse('package:flutter_fjs/'));
  if (resolved != null && resolved.scheme == 'file') {
    // package URIs point at the package's lib/ — the root is one up
    return Directory(resolved.toFilePath()).parent.absolute.path;
  }
  var dir = Directory.current;
  while (true) {
    final cfg = File('${dir.path}/.dart_tool/package_config.json');
    if (cfg.existsSync()) {
      final map = jsonDecode(cfg.readAsStringSync()) as Map<String, dynamic>;
      for (final p in (map['packages'] as List<dynamic>).cast<Map<String, dynamic>>()) {
        if (p['name'] == 'flutter_fjs') {
          final rootUri = Uri.parse(p['rootUri'] as String);
          return Directory(cfg.parent.uri.resolveUri(rootUri).toFilePath()).absolute.path;
        }
      }
    }
    final parent = dir.parent;
    if (parent.path == dir.path) break;
    dir = parent;
  }
  throw const FileSystemException(
      'cannot locate the flutter_fjs package — run from a host that depends on it');
}
